<launch>
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="multi_robot_name" default="" />

  <!-- Work mode -->
  <arg name='work_mode' default="real" />

  <!-- Init TurtleBot3 pose -->
  <arg name="x_pos" default="-0.2"/>
  <arg name="y_pos" default="1.6"/>
  <arg name="z_pos" default="0"/>
  <arg name="yaw_pos" default="3.14"/>

  <!-- View range for obstacle avoidance -->
  <arg name="view_range" default="[30, 75]" />

  <!-- PID controller parameters -->
  <arg name="Kp" default="0.2"/>
  <arg name="Ki" default="0.008"/>
  <arg name="Kd" default="0.2" />

  <!-- Line color -->
  # <arg name="lower_HSV" default="[20, 100, 100]" />
  # <arg name="upper_HSV" default="[50, 255, 255]" />
  <arg name="lower_LAB" default="[0, -10, 10]" />
  <arg name="upper_LAB" default="[90, 10, 100]" />
  
  <!-- Line center shift -->
  <arg name="line_center_shift" default="0" />

  <!-- Stop sign threshold -->
  
  <!-- TODO: Change the area threshold and stop delay for the real world -->
  <arg name="area_threshold" default="3300" />
  <arg name="stop_delay" default="18" />

  <!-- April tag family -->
  <arg name="tag_family" default="tag36h11" />
  <arg name="tag_area_threshold" default="10000" />

  <!-- Default linear speed -->
  <arg name="linear_x" default="0.1"/>

  <!-- Load Burger description -->
  <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />


  <!-- Obstacle Avoidance node -->
  <node pkg="aue_finals" type="obstacle_avoidance.py" name="obstacle_avoidance">
    <param name="work_mode" value="$(arg work_mode)" />
    <param name="linear_x" value="$(arg linear_x)" />
    <param name="view_range" value="$(arg view_range)" />
  </node>

  <!-- Line Detection node -->
  <node pkg="aue_finals" type="line_detection.py" name="line_detection">
    <param name="work_mode" value="$(arg work_mode)" />
    <param name="Kp" value="$(arg Kp)"/>
    <param name="Ki" value="$(arg Ki)"/>
    <param name="Kd" value="$(arg Kd)" />
    <param name="lower_LAB" value="$(arg lower_LAB)" />
    <param name="upper_LAB" value="$(arg upper_LAB)" />
    <param name="line_center_shift" value="$(arg line_center_shift)" />
  </node>

  <!-- Stop sign detection node -->
  <node pkg="aue_finals" type="stop_sign_detection_cv.py" name="stop_sign_detection">
    <param name="work_mode" value="$(arg work_mode)" />
  </node>

  <!-- april tag detection node -->
  <node pkg="aue_finals" type="apriltag_detection.py" name="apriltag_detection">
    <param name="work_mode" value="$(arg work_mode)" />
    <param name="linear_x" value="$(arg linear_x)" />
    <param name="tag_family" value="$(arg tag_family)" />
    <param name="tag_area_threshold" value="$(arg tag_area_threshold)" />
  </node>

  <!-- Main controller -->
  <node pkg="aue_finals" type="main_controller.py" name="main_controller">
    <param name="work_mode" value="$(arg work_mode)" />
    <param name="linear_x" value="$(arg linear_x)" />
    <param name="area_threshold" value="$(arg area_threshold)" />
    <param name="stop_delay" value="$(arg stop_delay)" />
  </node>

</launch>
